Dockerfile.rocm

```diff
@@ -105,6 +110,26 @@ RUN cd /opt/rocm/share/amd_smi \
 FROM scratch AS export_amdsmi
 COPY --from=build_amdsmi /opt/rocm/share/amd_smi/dist/*.whl /
 
+FROM base as build_pytorch
+ARG PYTORCH_BRANCH="v2.5.0-rc1"
+ARG PYTORCH_VISION_BRANCH="v1.19.1"
+ARG PYTORCH_REPO="https://github.com/pytorch/pytorch.git"
+ARG PYTORCH_VISION_REPO="https://github.com/pytorch/vision.git"
+RUN git clone --branch ${PYTORCH_BRANCH} --depth 1 ${PYTORCH_REPO} pytorch \
+    && cd pytorch \
+    && python tools/amd_build/build_amd.py \
+    && CMAKE_PREFIX_PATH=$(python3 -c 'import sys; print(sys.prefix)') python3 setup.py bdist_wheel --dist-dir=dist \
+    && cd .. \
+    && git clone --branch ${PYTORCH_VISION_BRANCH} --depth 1 ${PYTORCH_VISION_REPO} vision \
+    && cd vision \
+    && python3 setup.py bdist_wheel --dist-dir=dist
+FROM scratch as export_pytorch_1
+ARG COMMON_WORKDIR
+COPY --from=build_pytorch ${COMMON_WORKDIR}/pytorch/dist/*.whl /
+COPY --from=build_pytorch ${COMMON_WORKDIR}/vision/dist/*.whl /
+FROM scratch as export_pytorch_0
+from export_pytorch_${BUILD_PYTORCH} as export_pytorch
```

Comment:
A couple points here:

1. Updating PyTorch to 2.5.0 is necessary in any case for stateless device count and respecting CUDA_VISIBLE_DEVICES, both of which are needed for tests

2. Building PyTorch adds an extra half hour to our build in the best case. If `BUILD_PYTHON=0`, we should install PyTorch wheels instead. Of course, if we are going to update Torch unconditionally, we might want to change the arg name in that case, e.g. `BUILD_PYTORCH_FROM_SOURCE`. And maybe also add in args like `PYTORCH_WHEEL_NAME` and `VISION_WHEEL_NAME` for use in that case.
- If the flag that says build PyTorch from source is 0, then instead of building wheels, we should download the wheels  instead. This can be done say by setting `export_pytorch_0` to be the following:
```
FROM base AS export_pytorch_0
ARG PYTORCH_WHEEL_NAME
ARG VISION_WHEEL_NAME
RUN mkdir -p pytorch/dist \
    && case "$(ls /opt | grep -Po 'rocm-[0-9]\.[0-9]')" in \
           *"rocm-6.0"*) \
               export WHL_URL=https://download.pytorch.org/whl/rocm6.0;;
           *"rocm-6.1"*) \
               export WHL_URL=https://download.pytorch.org/whl/nightly/rocm6.1;; \
           *"rocm-6.2"*) \
               export WHL_URL=https://download.pytorch.org/whl/nightly/rocm6.2;; \
           *) ;; esac \
    && python3 -m pip download --pre torch==${PYTORCH_WHEEL_NAME} torchvision==${VISION_WHEEL_NAME}  --index-url ${WHL_URL}  -d pytorch/dist
```

Reply:
As last couple weeks showed us, the full whls are unusable, so until we get an official lightweight whl, or a ROCm release with aversion that works for us, there are the following scenarios we are going to support:
1. Your image has the versions you want to use. Don't build anything.
2. Build the pinned versions with just the right sets of dependencies, so it works out of the box.

What we may need to do is build torch on top of the installed hipblaslt, because there seem to be API differences between .8 and .10, but that will prevent from running them in parallel, so I'd rather run a few more tests before deciding.

Reply:
1. Can you elaborate on how the full wheels are unusable? They seem to be working for me.
2. Strictly speaking, PyTorch depends on hipBLASLt and RCCL. So not installing hipBLASLt before PyTorch is risky. I think a reasonable contract we can offer is that we'll advance the (hipBLASLt, RCCL) versions only if they don't break APIs sufficiently that we have to build+install any of them before we build Torch.